File: sponge-jython/examples/script/py/knowledge_base_auto_enable.py
"""
Sponge Knowledge base
Auto-enable
"""

from java.util.concurrent.atomic import AtomicInteger

def onInit():
    # Variables for assertions only
    sponge.setVariable("counter", AtomicInteger(0))

class AutoAction(Action):
    def onCall(self):
        self.logger.debug("Running")
        sponge.getVariable("counter").incrementAndGet()
        return None

class AutoFilter(Filter):
    def onConfigure(self):
        self.event = "e1"
    def onAccept(self, event):
        self.logger.debug("Received event: {}", event.name)
        sponge.getVariable("counter").incrementAndGet()
        return True

class AutoTrigger(Trigger):
    def onConfigure(self):
        self.event = "e1"
    def onRun(self, event):
        self.logger.debug("Received event: {}", event.name)
        sponge.getVariable("counter").incrementAndGet()

class AutoRule(Rule):
    def onConfigure(self):
        self.events = ["e1", "e2"]
    def onRun(self, event):
        self.logger.debug("Running for sequence: {}", self.eventSequence)
        sponge.getVariable("counter").incrementAndGet()

class AutoCorrelator(Correlator):
    def onConfigure(self):
        self.events = ["e1", "e2"]
    def onAcceptAsFirst(self, event):
        return event.name == "e1"
    def onEvent(self, event):
        self.logger.debug("Received event: {}", event.name)
        if event.name == "e2":
            sponge.getVariable("counter").incrementAndGet()
            self.finish()

def onStartup():
    sponge.call("AutoAction")
    sponge.event("e1").send()
    sponge.event("e2").send()

File: sponge-kotlin/src/test/kotlin/examples/KnowledgeBaseAutoEnable.kt
/*
 * Sponge Knowledge Base
 * Auto-enable
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.event.Event
import org.openksavi.sponge.kotlin.KAction
import org.openksavi.sponge.kotlin.KCorrelator
import org.openksavi.sponge.kotlin.KFilter
import org.openksavi.sponge.kotlin.KKnowledgeBase
import org.openksavi.sponge.kotlin.KRule
import org.openksavi.sponge.kotlin.KTrigger
import java.util.concurrent.atomic.AtomicInteger

class KnowledgeBaseAutoEnable : KKnowledgeBase() {

    override fun onInit() {
        // Variables for assertions only
        sponge.setVariable("counter", AtomicInteger(0))
    }

    class AutoAction : KAction() {
        fun onCall(): Any? {
            logger.debug("Running")
            sponge.getVariable<AtomicInteger>("counter").incrementAndGet()
            return null
        }
    }

    class AutoFilter : KFilter() {
        override fun onConfigure() {
            withEvent("e1")
        }
        override fun onAccept(event: Event): Boolean {
            logger.debug("Received event: {}", event.name)
            sponge.getVariable<AtomicInteger>("counter").incrementAndGet()
            return true
        }
    }

    class AutoTrigger : KTrigger() {
        override fun onConfigure() {
            withEvent("e1")
        }
        override fun onRun(event: Event) {
            logger.debug("Received event: {}", event.name)
            sponge.getVariable<AtomicInteger>("counter").incrementAndGet()
        }
    }

    class AutoRule : KRule() {
        override fun onConfigure() {
            withEvents("e1", "e2")
        }
        override fun onRun(event: Event) {
            logger.debug("Running for sequence: {}", eventSequence)
            sponge.getVariable<AtomicInteger>("counter").incrementAndGet()
        }
    }

    class AutoCorrelator : KCorrelator() {
        override fun onConfigure() {
            withEvents("e1", "e2")
        }

        override fun onAcceptAsFirst(event: Event) = event.name == "e1"

        override fun onEvent(event: Event) {
            logger.debug("Received event: {}", event.name)
            if (event.name == "e2") {
                sponge.getVariable<AtomicInteger>("counter").incrementAndGet()
                finish()
            }
        }
    }

    override fun onStartup() {
        sponge.call("AutoAction")
        sponge.event("e1").send()
        sponge.event("e2").send()
    }
}
